File: 04_implementation/frontend/src/api/apus.ts
import { request } from "./client";

export type Apu = {
  row_number: number;
  codigo_apu: string;
  descripcion_apu: string;
  unidad: string;
  costo_total_formula: string | null;
};

export type ApuSearchResponse = {
  items: Apu[];
  total: number;
  search: string;
  can_create: boolean;
  suggested_code: string | null;
};

export type ApuItemType = "MATERIAL" | "LABOR" | "CUADRILLA";

export type ApuDetailPayload = {
  tipo_item: ApuItemType;
  codigo_item: string;
  cantidad: string;
  source_row_number?: number | null;
  cantidad_formula?: string | null;
};

export type ApuDraftDetail = ApuDetailPayload & {
  row_number: number;
  source_row_number: number;
  cantidad_formula: string | null;
  descripcion: string | null;
  und: string | null;
  costo: string | null;
};

export type ApuDraftFromBase = {
  base_row_number: number;
  base_codigo_apu: string;
  codigo_apu: string;
  descripcion_apu: string;
  unidad: string;
  detalles: ApuDraftDetail[];
};

export type ApuCreatePayload = {
  codigo_apu: string;
  descripcion_apu: string;
  unidad: string;
  detalles: ApuDetailPayload[];
};

export type ApuCreateResponse = {
  header_row: number;
  detail_rows: number;
  backup_path: string;
  formula_integrity_ok: boolean;
};

export function searchApus(search: string): Promise<ApuSearchResponse> {
  const query = new URLSearchParams({ search, limit: "80" });
  return request<ApuSearchResponse>(`/apus?${query.toString()}`);
}

export function suggestApuCode(description: string): Promise<string> {
  const query = new URLSearchParams({ description });
  return request<{ suggested_code: string }>(
    `/apus/suggest-code?${query.toString()}`,
  ).then((response) => response.suggested_code);
}

export function createApu(payload: ApuCreatePayload): Promise<ApuCreateResponse> {
  return request<ApuCreateResponse>("/apus", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function getApuDraft(rowNumber: number): Promise<ApuDraftFromBase> {
  return request<ApuDraftFromBase>(`/apus/${rowNumber}/draft`);
}

File: 04_implementation/backend/app/schemas/apus.py
from __future__ import annotations

from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, Field


ItemType = Literal["MATERIAL", "LABOR", "CUADRILLA"]


class ApuDetail(BaseModel):
    tipo_item: ItemType
    codigo_item: str = Field(..., min_length=1)
    cantidad: Decimal | str = Decimal("1")
    source_row_number: Optional[int] = None
    cantidad_formula: Optional[str] = None


class ApuHeader(BaseModel):
    codigo_apu: str = Field(..., min_length=1)
    descripcion_apu: str = Field(..., min_length=1)
    unidad: str = Field(..., min_length=1)


class ApuRead(ApuHeader):
    row_number: int
    costo_total_formula: Optional[str] = None


class ApuCreate(ApuHeader):
    detalles: list[ApuDetail] = Field(default_factory=list)


class ApuDraftDetail(BaseModel):
    row_number: int
    source_row_number: int
    tipo_item: ItemType
    codigo_item: str = Field(..., min_length=1)
    cantidad: str = "1"
    cantidad_formula: Optional[str] = None
    descripcion: Optional[str] = None
    und: Optional[str] = None
    costo: Optional[str] = None


class ApuDraftRead(BaseModel):
    base_row_number: int
    base_codigo_apu: str
    codigo_apu: str = ""
    descripcion_apu: str = Field(..., min_length=1)
    unidad: str = Field(..., min_length=1)
    detalles: list[ApuDraftDetail] = Field(default_factory=list)
